internal/engine/exit/controller.go

```diff
@@ -2,125 +2,203 @@ package exit
+// Controller handles normal process termination. Either Tilt completed all its work,
 // or it determined that it was unable to complete the work it was assigned.
 type Controller struct {
-}
+	pid       int64
+	startTime time.Time
+	client    ctrlclient.Client
 
-func NewController() *Controller {
-	return &Controller{}
+	mu      sync.Mutex
+	tiltRun *tiltruns.TiltRun
 }
 
-func (c *Controller) shouldExit(store store.RStore) Action {
-	state := store.RLockState()
-	defer store.RUnlockState()
+var _ store.Subscriber = &Controller{}
 
-	// If state already has an ExitSignal or engine is NOT in CI mode (i.e. it's in interactive "Up" mode),
-	// there's nothing to do
-	if state.ExitSignal || !state.EngineMode.IsCIMode() {
-		return Action{}
+func NewController(cli ctrlclient.Client) *Controller {
+	return &Controller{
+		pid:       int64(os.Getpid()),
+		startTime: time.Now(),
+		client:    cli,
 	}
+}
 
-	// If the tiltfile failed, exit immediately.
-	err := state.TiltfileState.LastBuild().Error
-	if err != nil {
-		return Action{ExitSignal: true, ExitError: err}
+func (c *Controller) OnChange(ctx context.Context, st store.RStore, summary store.ChangeSummary) {
+	if summary.IsLogOnly() {
+		return
 	}
 
-	// If any of the individual builds failed, exit immediately.
-	for _, mt := range state.ManifestTargets {
-		err := mt.State.LastBuild().Error
-		if err != nil {
-			return Action{ExitSignal: true, ExitError: err}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.tiltRun == nil {
+		if initialized, err := c.initialize(ctx, st); err != nil {
+			st.Dispatch(store.NewErrorAction(fmt.Errorf("failed to initialize ExitController: %v", err)))
+			return
+		} else if !initialized {
+			// engine is still starting up, no-op until ready for initialization
+			return
 		}
 	}
 
-	// Check the runtime state of all resources.
-	// If any of the resources are in error, exit.
-	allOK := true
+	newStatus := c.makeLatestStatus(st)
+	if err := c.handleLatestStatus(ctx, st, newStatus); err != nil {
+		logger.Get(ctx).Debugf("failed to update TiltRun status: %v", err)
+	}
+}
+
+func (c *Controller) initialize(ctx context.Context, st store.RStore) (bool, error) {
+	tiltRun := c.makeTiltRun(st)
+	if tiltRun == nil {
+		return false, nil
+	}
+
+	// TODO(milas): rather than implicitly creating the TiltRun object here, it should
+	// 	be created explicitly as part of loading the Tiltfile
+	if err := c.client.Create(ctx, tiltRun); err != nil {
+		return false, fmt.Errorf("failed to create TiltRun API object: %v", err)
+	}
+
+	c.tiltRun = tiltRun
+
+	return true, nil
+}
+
+func (c *Controller) makeTiltRun(st store.RStore) *tiltruns.TiltRun {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	// engine hasn't finished initialization - Tiltfile hasn't been loaded yet
+	if state.TiltfilePath == "" {
+		return nil
+	}
+
+	tiltRun := &tiltruns.TiltRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "Tiltfile",
```

Comment:
I totally punted on this for the moment since we can only have 1x run/Tiltfile per process currently, but yep - it will likely* need to get a unique name in the future.

Since `/` isn't an allowed character for entity names (and `Users-milas-dev-tilt-web-storybook.tiltfile` is an obnoxious name anyway), I think we might want to do something like `docker-compose` in the future where we pick the directory name but maybe with K8s flavor added in with some randomness at the end?

* But also - apiserver supports namespaces! So that might be a logical boundary anyway in which case you wouldn't conflict! But even in that case, I could see a `--attach` type functionality to merge them into 1 session 🤔 

TL;DR It's fine for now 🤷